src/dist/replication/lib/replica_restore.cpp

```diff
@@ -424,42 +324,36 @@ dsn::error_code replica::restore_checkpoint()
         skip_bad_partition = true;
     }
 
-    // then create a local restore dir
-    std::ostringstream os;
-    os << _dir << "/restore." << restore_req.policy_name << "." << restore_req.time_stamp;
-    std::string restore_dir = os.str();
+    // then create a local restore dir if it doesn't exist
     if (!utils::filesystem::directory_exists(_dir) && !utils::filesystem::create_directory(_dir)) {
         derror("create dir %s failed", _dir.c_str());
-        return dsn::ERR_FILE_OPERATION_FAILED;
+        return ERR_FILE_OPERATION_FAILED;
     }
+
     // we don't remove the old restore.policy_name.backup_id
+    std::ostringstream os;
+    os << _dir << "/restore." << restore_req.policy_name << "." << restore_req.time_stamp;
+    std::string restore_dir = os.str();
     if (!utils::filesystem::directory_exists(restore_dir) &&
         !utils::filesystem::create_directory(restore_dir)) {
         derror("create dir %s failed", restore_dir.c_str());
-        return dsn::ERR_FILE_OPERATION_FAILED;
+        return ERR_FILE_OPERATION_FAILED;
     }
 
-    // then find a valid checkpoint dir to copy
+    // then find a valid checkpoint dir and download it
     std::string remote_chkpt_dir;
-    dsn::error_code err = find_valid_checkpoint(restore_req, remote_chkpt_dir);
-
-    if (err == dsn::ERR_OK) {
+    error_code err = find_valid_checkpoint(restore_req, remote_chkpt_dir);
+    if (err == ERR_OK) {
         err = download_checkpoint(restore_req, remote_chkpt_dir, restore_dir);
-        if (_restore_status == ERR_CORRUPTION) {
-            if (skip_bad_partition) {
-                err = skip_restore_partition(restore_dir);
-            } else {
-                tell_meta_to_restore_rollback();
-                return ERR_CORRUPTION;
-            }
-        }
-    } else if (err == ERR_OBJECT_NOT_FOUND) { // find valid checkpoint failed
+    }
+
+    if (err == ERR_OBJECT_NOT_FOUND || err == ERR_CORRUPTION) {
```

Comment:
This condition will handle `find_valid_checkpoint` and `download_checkpoint`'s error `ERR_OBJECT_NOT_FOUND` and `ERR_CORRUPTION`, will those two functions return other error code?

Reply:
yes, if return other error code, the restore_checkpoint will return the error code, and wil retry later. Only these `ERR_OBJECT_NOT_FOUND` and `ERR_CORRUPTION` should two judge whether they allow `skip_bad_partition` or not. if it is not allowed, the restore will rollback.